From f6bfadf6fdf0eb22b5a09d884aaa85cd56b22100 Mon Sep 17 00:00:00 2001
From: Chen Li1 <li1.chen@intel.com>
Date: Sun, 19 Jan 2020 17:36:37 +0800
Subject: [PATCH] Changes for webrtc upgrade

---
 erizo/src/erizo/MediaStream.cpp | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/erizo/src/erizo/MediaStream.cpp b/erizo/src/erizo/MediaStream.cpp
index 4327c6f..00e046a 100644
--- a/erizo/src/erizo/MediaStream.cpp
+++ b/erizo/src/erizo/MediaStream.cpp
@@ -21,7 +21,7 @@
 #include "rtp/RtcpForwarder.h"
 // #include "rtp/RtpSlideShowHandler.h"
 // #include "rtp/RtpTrackMuteHandler.h"
-#include "rtp/BandwidthEstimationHandler.h"
+// #include "rtp/BandwidthEstimationHandler.h"
 // #include "rtp/FecReceiverHandler.h"
 #include "rtp/RtcpProcessorHandler.h"
 // #include "rtp/RtpRetransmissionHandler.h"
@@ -366,7 +366,7 @@ void MediaStream::initializePipeline() {
   // pipeline_->addFront(std::make_shared<RtpPaddingGeneratorHandler>());
   // pipeline_->addFront(std::make_shared<PliPacerHandler>());
   if (simulcast_ && is_publisher_) {
-    pipeline_->addFront(std::make_shared<BandwidthEstimationHandler>());
+    // pipeline_->addFront(std::make_shared<BandwidthEstimationHandler>());
   }
   // pipeline_->addFront(std::make_shared<RtpPaddingRemovalHandler>());
   // pipeline_->addFront(std::make_shared<RtcpFeedbackGenerationHandler>());
@@ -410,10 +410,10 @@ int MediaStream::deliverFeedback_(std::shared_ptr<DataPacket> fb_packet) {
   }
   if (isVideoSourceSSRC(recvSSRC)) {
     fb_packet->type = VIDEO_PACKET;
-    sendPacketAsync(fb_packet);
+    sendPacketAsync(std::make_shared<DataPacket>(*fb_packet));
   } else if (isAudioSourceSSRC(recvSSRC)) {
     fb_packet->type = AUDIO_PACKET;
-    sendPacketAsync(fb_packet);
+    sendPacketAsync(std::make_shared<DataPacket>(*fb_packet));
   } else {
     ELOG_DEBUG("%s deliverFeedback unknownSSRC: %u, localVideoSSRC: %u, localAudioSSRC: %u",
                 toLog(), recvSSRC, this->getVideoSourceSSRC(), this->getAudioSourceSSRC());
@@ -563,6 +563,7 @@ void MediaStream::sendPLIToFeedback() {
       this->getVideoSourceSSRC()));
   }
 }
+
 // changes the outgoing payload type for in the given data packet
 void MediaStream::sendPacketAsync(std::shared_ptr<DataPacket> packet) {
   if (!sending_) {
-- 
2.7.4

