From 636830ed0ed15ff629490d897dd4acbb3fdb8abe Mon Sep 17 00:00:00 2001
From: Chen Li1 <li1.chen@intel.com>
Date: Tue, 12 May 2020 13:34:23 +0800
Subject: [PATCH] Fix thread issues when closing

---
 agent/agent.c     | 15 ++++++++++-----
 agent/component.c | 11 ++++++-----
 2 files changed, 16 insertions(+), 10 deletions(-)

diff --git a/agent/agent.c b/agent/agent.c
index 94b3ac8..67d1f55 100644
--- a/agent/agent.c
+++ b/agent/agent.c
@@ -2631,11 +2631,7 @@ nice_agent_dispose (GObject *object)
   GSList *i;
   NiceAgent *agent = NICE_AGENT (object);
 
-#if GLIB_CHECK_VERSION(2,31,8)
-  g_rec_mutex_clear (&agent->mutex);
-#else
-  g_static_rec_mutex_free (&agent->mutex);
-#endif
+  agent_lock(agent);
 
   /* step: free resources for the binding discovery timers */
   discovery_free (agent);
@@ -2690,6 +2686,15 @@ nice_agent_dispose (GObject *object)
     g_main_context_unref (agent->main_context);
   agent->main_context = NULL;
 
+  agent_unlock(agent);
+
+/* Can't clear mutex here due to libnice bug,
+ * nice_agent_g_source_cb may be called after this function */
+#if GLIB_CHECK_VERSION(2,31,8)
+  //g_rec_mutex_clear (&agent->mutex);
+#else
+  //g_static_rec_mutex_free (&agent->mutex);
+#endif
   if (G_OBJECT_CLASS (nice_agent_parent_class)->dispose)
     G_OBJECT_CLASS (nice_agent_parent_class)->dispose (object);
 
diff --git a/agent/component.c b/agent/component.c
index 2121e67..d5ac214 100644
--- a/agent/component.c
+++ b/agent/component.c
@@ -89,16 +89,17 @@ component_free (Component *cmp)
     nice_candidate_free (cmp->restart_candidate),
       cmp->restart_candidate = NULL;
 
-  for (i = cmp->sockets; i; i = i->next) {
-    NiceSocket *udpsocket = i->data;
-    nice_socket_free (udpsocket);
-  }
-
+  /* Destroy g_source before free socket */
   for (i = cmp->gsources; i; i = i->next) {
     GSource *source = i->data;
     g_source_destroy (source);
     g_source_unref (source);
   }
+
+  for (i = cmp->sockets; i; i = i->next) {
+    NiceSocket *udpsocket = i->data;
+    nice_socket_free (udpsocket);
+  }
  
   for (i = cmp->incoming_checks; i; i = i->next) {
     IncomingCheck *icheck = i->data;
-- 
2.7.4

