--- configure	2015-10-08 10:01:49.782029013 +0800
+++ configure	2015-10-08 10:02:44.426029185 +0800
@@ -731,11 +731,14 @@
     if pkg_cflags:
       output['include_dirs'] += (
           filter(None, map(str.strip, pkg_cflags.split('-I'))))
+    elif options.__dict__[shared_lib + '_includes']:
+      output['include_dirs'] += [options.__dict__[shared_lib + '_includes']]
 
     # libpath needs to be provided ahead libraries
     if pkg_libpath:
-      output['libraries'] += (
-          filter(None, map(str.strip, pkg_cflags.split('-L'))))
+      output['libraries'] += [pkg_libpath]
+    elif options.__dict__[shared_lib + '_libpath']:
+      output['libraries'] += ['-L%s' % options.__dict__[shared_lib + '_libpath']]
 
     default_libs = getattr(options, shared_lib + '_libname')
     default_libs = map('-l{0}'.format, default_libs.split(','))
