--- configure	2015-10-09 08:57:22.890289207 +0800
+++ configure	2015-10-09 09:04:30.862290557 +0800
@@ -165,6 +165,27 @@
     dest='shared_openssl_libpath',
     help='a directory to search for the shared OpenSSL DLLs')
 
+shared_optgroup.add_option('--shared-tcmalloc',
+    action='store_true',
+    dest='shared_tcmalloc',
+    help='Link to a shared tcmalloc DLL')
+
+shared_optgroup.add_option('--shared-tcmalloc-includes',
+    action='store',
+    dest='shared_tcmalloc_includes',
+    help='Directory containing tcmalloc header files')
+
+shared_optgroup.add_option('--shared-tcmalloc-libpath',
+    action='store',
+    dest='shared_tcmalloc_libpath',
+    help='A directory to search for the shared tcmalloc DLL')
+
+shared_optgroup.add_option('--shared-tcmalloc-libname',
+    action='store',
+    dest='shared_tcmalloc_libname',
+    default='tcmalloc,dl',
+    help='alternative lib name to link to [default: %default]')
+
 shared_optgroup.add_option('--shared-zlib',
     action='store_true',
     dest='shared_zlib',
@@ -1069,6 +1090,7 @@
 configure_library('zlib', output)
 configure_library('http_parser', output)
 configure_library('libuv', output)
+configure_library('tcmalloc', output)
 configure_v8(output)
 configure_openssl(output)
 configure_winsdk(output)
