Index: src/third_party/opus/opus.gyp
===================================================================
--- src/third_party/opus/opus.gyp	(revision 266564)
+++ src/third_party/opus/opus.gyp	(working copy)
@@ -59,7 +59,7 @@
           # Suppress a warning given by opus_decoder.c that tells us
           # optimizations are turned off.
           'cflags': [
-            '-Wno-#pragma-messages',
+            # '-Wno-#pragma-messages',
           ],
           'xcode_settings': {
             'WARNING_CFLAGS': [
