diff -rc webrtc.upstream/modules/modules.gyp webrtc/modules/modules.gyp
*** webrtc.upstream/modules/modules.gyp	2014-05-01 02:50:27.668761000 +0800
--- webrtc/modules/modules.gyp	2014-09-24 15:31:01.238619080 +0800
***************
*** 20,29 ****
      'audio_coding/neteq/neteq.gypi',
      'audio_coding/neteq4/neteq.gypi',
      'audio_conference_mixer/source/audio_conference_mixer.gypi',
!     'audio_device/audio_device.gypi',
      'audio_processing/audio_processing.gypi',
      'bitrate_controller/bitrate_controller.gypi',
!     'desktop_capture/desktop_capture.gypi',
      'media_file/source/media_file.gypi',
      'pacing/pacing.gypi',
      'remote_bitrate_estimator/remote_bitrate_estimator.gypi',
--- 20,29 ----
      'audio_coding/neteq/neteq.gypi',
      'audio_coding/neteq4/neteq.gypi',
      'audio_conference_mixer/source/audio_conference_mixer.gypi',
! #    'audio_device/audio_device.gypi',
      'audio_processing/audio_processing.gypi',
      'bitrate_controller/bitrate_controller.gypi',
! #    'desktop_capture/desktop_capture.gypi',
      'media_file/source/media_file.gypi',
      'pacing/pacing.gypi',
      'remote_bitrate_estimator/remote_bitrate_estimator.gypi',
***************
*** 31,39 ****
      'utility/source/utility.gypi',
      'video_coding/codecs/i420/main/source/i420.gypi',
      'video_coding/main/source/video_coding.gypi',
!     'video_capture/video_capture.gypi',
      'video_processing/main/source/video_processing.gypi',
!     'video_render/video_render.gypi',
    ],
    'conditions': [
      ['include_opus==1', {
--- 31,39 ----
      'utility/source/utility.gypi',
      'video_coding/codecs/i420/main/source/i420.gypi',
      'video_coding/main/source/video_coding.gypi',
!  #   'video_capture/video_capture.gypi',
      'video_processing/main/source/video_processing.gypi',
!  #   'video_render/video_render.gypi',
    ],
    'conditions': [
      ['include_opus==1', {
diff -rc webrtc.upstream/modules/utility/source/utility.gypi webrtc/modules/utility/source/utility.gypi
*** webrtc.upstream/modules/utility/source/utility.gypi	2013-12-10 00:23:00.328294000 +0800
--- webrtc/modules/utility/source/utility.gypi	2014-09-24 15:31:16.690618545 +0800
***************
*** 31,37 ****
          'file_player_impl.h',
          'file_recorder_impl.cc',
          'file_recorder_impl.h',
!         'helpers_android.cc',
          'process_thread_impl.cc',
          'process_thread_impl.h',
          'rtp_dump_impl.cc',
--- 31,37 ----
          'file_player_impl.h',
          'file_recorder_impl.cc',
          'file_recorder_impl.h',
! #        'helpers_android.cc',
          'process_thread_impl.cc',
          'process_thread_impl.h',
          'rtp_dump_impl.cc',
diff -rc webrtc.upstream/system_wrappers/source/system_wrappers.gyp webrtc/system_wrappers/source/system_wrappers.gyp
*** webrtc.upstream/system_wrappers/source/system_wrappers.gyp	2014-02-06 02:43:46.255322000 +0800
--- webrtc/system_wrappers/source/system_wrappers.gyp	2014-09-24 15:31:42.766617641 +0800
***************
*** 144,149 ****
--- 144,167 ----
            'link_settings': {
              'libraries': [ '-lrt', ],
            },
+           'sources!': [
+             'atomic32_mac.cc',
+             'atomic32_win.cc',
+             'condition_variable_event_win.cc',
+             'condition_variable_event_win.h',
+             'condition_variable_native_win.cc',
+             'condition_variable_native_win.h',
+             'critical_section_win.cc',
+             'critical_section_win.h',
+             'event_win.cc',
+             'event_win.h',
+             'rw_lock_win.cc',
+             'rw_lock_win.h',
+             'thread_win.cc',
+             'thread_win.h',
+             'trace_win.cc',
+             'trace_win.h',
+           ],
          }],
          ['OS=="mac"', {
            'link_settings': {
diff -rc webrtc.upstream/webrtc.gyp webrtc/webrtc.gyp
*** webrtc.upstream/webrtc.gyp	2013-12-19 04:28:25.711986000 +0800
--- webrtc/webrtc.gyp	2014-09-24 15:31:53.150617283 +0800
***************
*** 6,79 ****
  # in the file PATENTS.  All contributing project authors may
  # be found in the AUTHORS file in the root of the source tree.
  {
-   'conditions': [
-     ['include_tests==1', {
-       'includes': [
-         'webrtc_tests.gypi',
-       ],
-     }],
-   ],
    'includes': [
      'build/common.gypi',
-     'video/webrtc_video.gypi',
    ],
-   'variables': {
-     'webrtc_all_dependencies': [
-       'common_audio/common_audio.gyp:*',
-       'common_video/common_video.gyp:*',
-       'modules/modules.gyp:*',
-       'system_wrappers/source/system_wrappers.gyp:*',
-       'video_engine/video_engine.gyp:*',
-       'voice_engine/voice_engine.gyp:*',
-       '<(webrtc_vp8_dir)/vp8.gyp:*',
-     ],
-   },
    'targets': [
      {
        'target_name': 'webrtc_all',
        'type': 'none',
        'dependencies': [
!         '<@(webrtc_all_dependencies)',
!         'webrtc',
!       ],
!       'conditions': [
!         ['include_tests==1', {
!           'dependencies': [
!             'common_video/common_video_unittests.gyp:*',
!             'system_wrappers/source/system_wrappers_tests.gyp:*',
!             'test/metrics.gyp:*',
!             'test/test.gyp:*',
!             'test/webrtc_test_common.gyp:webrtc_test_common_unittests',
!             'tools/tools.gyp:*',
!             'webrtc_tests',
!           ],
!         }],
!         ['build_with_chromium==0 and OS=="android"', {
!           'dependencies': [
!             '../tools/android/android_tools_precompiled.gyp:*',
!           ],
!         }],
!       ],
!     },
!     {
!       # TODO(pbos): This is intended to contain audio parts as well as soon as
!       #             VoiceEngine moves to the same new API format.
!       'target_name': 'webrtc',
!       'type': 'static_library',
!       'sources': [
!         'call.h',
!         'config.h',
!         'experiments.h',
!         'frame_callback.h',
!         'transport.h',
!         'video_receive_stream.h',
!         'video_renderer.h',
!         'video_send_stream.h',
! 
!         '<@(webrtc_video_sources)',
!       ],
!       'dependencies': [
!         '<@(webrtc_video_dependencies)',
        ],
      },
    ],
--- 6,20 ----
  # in the file PATENTS.  All contributing project authors may
  # be found in the AUTHORS file in the root of the source tree.
  {
    'includes': [
      'build/common.gypi',
    ],
    'targets': [
      {
        'target_name': 'webrtc_all',
        'type': 'none',
        'dependencies': [
!         'modules/modules.gyp:*',
        ],
      },
    ],
