Index: src/webrtc/webrtc.gyp
===================================================================
--- src/webrtc/webrtc.gyp	(revision 7370)
+++ src/webrtc/webrtc.gyp	(working copy)
@@ -6,75 +6,17 @@
 # in the file PATENTS.  All contributing project authors may
 # be found in the AUTHORS file in the root of the source tree.
 {
-  'conditions': [
-    ['include_tests==1', {
-      'includes': [
-        'webrtc_tests.gypi',
-      ],
-    }],
-  ],
   'includes': [
     'build/common.gypi',
-    'video/webrtc_video.gypi',
   ],
-  'variables': {
-    'webrtc_all_dependencies': [
-      'common_audio/common_audio.gyp:*',
-      'common_video/common_video.gyp:*',
-      'modules/modules.gyp:*',
-      'system_wrappers/source/system_wrappers.gyp:*',
-      'video_engine/video_engine.gyp:*',
-      'voice_engine/voice_engine.gyp:*',
-      '<(webrtc_vp8_dir)/vp8.gyp:*',
-    ],
-  },
   'targets': [
     {
       'target_name': 'webrtc_all',
       'type': 'none',
       'dependencies': [
-        '<@(webrtc_all_dependencies)',
-        'webrtc',
+        'modules/modules.gyp:*',
+        'voice_engine/voice_engine.gyp:*',
       ],
-      'conditions': [
-        ['include_tests==1', {
-          'dependencies': [
-            'common_video/common_video_unittests.gyp:*',
-            'system_wrappers/source/system_wrappers_tests.gyp:*',
-            'test/metrics.gyp:*',
-            'test/test.gyp:*',
-            'test/webrtc_test_common.gyp:webrtc_test_common_unittests',
-            'tools/tools.gyp:*',
-            'webrtc_tests',
-          ],
-        }],
-        ['build_with_chromium==0 and OS=="android"', {
-          'dependencies': [
-            '../tools/android/android_tools_precompiled.gyp:*',
-          ],
-        }],
-      ],
     },
-    {
-      # TODO(pbos): This is intended to contain audio parts as well as soon as
-      #             VoiceEngine moves to the same new API format.
-      'target_name': 'webrtc',
-      'type': 'static_library',
-      'sources': [
-        'call.h',
-        'config.h',
-        'experiments.h',
-        'frame_callback.h',
-        'transport.h',
-        'video_receive_stream.h',
-        'video_renderer.h',
-        'video_send_stream.h',
-
-        '<@(webrtc_video_sources)',
-      ],
-      'dependencies': [
-        '<@(webrtc_video_dependencies)',
-      ],
-    },
   ],
 }
Index: src/webrtc/modules/utility/source/utility.gypi
===================================================================
--- src/webrtc/modules/utility/source/utility.gypi	(revision 7370)
+++ src/webrtc/modules/utility/source/utility.gypi	(working copy)
@@ -21,7 +21,6 @@
         '../interface/audio_frame_operations.h',
         '../interface/file_player.h',
         '../interface/file_recorder.h',
-        '../interface/helpers_android.h',
         '../interface/process_thread.h',
         '../interface/rtp_dump.h',
         'audio_frame_operations.cc',
@@ -31,13 +30,18 @@
         'file_player_impl.h',
         'file_recorder_impl.cc',
         'file_recorder_impl.h',
-        'helpers_android.cc',
         'process_thread_impl.cc',
         'process_thread_impl.h',
         'rtp_dump_impl.cc',
         'rtp_dump_impl.h',
       ],
       'conditions': [
+        ['OS=="android"', {
+          'sources': [
+            '../interface/helpers_android.h',
+            'helpers_android.cc',
+          ],
+        }],
         ['enable_video==1', {
           'dependencies': [
             'webrtc_video_coding',
Index: src/webrtc/modules/modules.gyp
===================================================================
--- src/webrtc/modules/modules.gyp	(revision 7370)
+++ src/webrtc/modules/modules.gyp	(working copy)
@@ -23,7 +23,6 @@
     'audio_device/audio_device.gypi',
     'audio_processing/audio_processing.gypi',
     'bitrate_controller/bitrate_controller.gypi',
-    'desktop_capture/desktop_capture.gypi',
     'media_file/source/media_file.gypi',
     'pacing/pacing.gypi',
     'remote_bitrate_estimator/remote_bitrate_estimator.gypi',
@@ -31,11 +30,18 @@
     'utility/source/utility.gypi',
     'video_coding/codecs/i420/main/source/i420.gypi',
     'video_coding/main/source/video_coding.gypi',
-    'video_capture/video_capture.gypi',
     'video_processing/main/source/video_processing.gypi',
-    'video_render/video_render.gypi',
   ],
   'conditions': [
+    ['include_capture==1', {
+      'includes': [
+        'desktop_capture/desktop_capture.gypi',
+        'video_capture/video_capture.gypi',
+      ],
+    }],
+    ['include_render==1', {
+      'includes': ['video_render/video_render.gypi',],
+    }],
     ['include_opus==1', {
       'includes': ['audio_coding/codecs/opus/opus.gypi',],
     }],
Index: src/webrtc/build/common.gypi
===================================================================
--- src/webrtc/build/common.gypi	(revision 7370)
+++ src/webrtc/build/common.gypi	(working copy)
@@ -55,6 +55,9 @@
     'include_opus%': '<(include_opus)',
     'rbe_components_path%': '<(rbe_components_path)',
 
+    'include_capture%': 1,
+    'include_render%': 1,
+
     # The Chromium common.gypi we use treats all gyp files without
     # chromium_code==1 as third party code. This disables many of the
     # preferred warning settings.
Index: src/webrtc/system_wrappers/source/system_wrappers.gyp
===================================================================
--- src/webrtc/system_wrappers/source/system_wrappers.gyp	(revision 7370)
+++ src/webrtc/system_wrappers/source/system_wrappers.gyp	(working copy)
@@ -144,6 +144,24 @@
           'link_settings': {
             'libraries': [ '-lrt', ],
           },
+          'sources!': [
+            'atomic32_mac.cc',
+            'atomic32_win.cc',
+            'condition_variable_event_win.cc',
+            'condition_variable_event_win.h',
+            'condition_variable_native_win.cc',
+            'condition_variable_native_win.h',
+            'critical_section_win.cc',
+            'critical_section_win.h',
+            'event_win.cc',
+            'event_win.h',
+            'rw_lock_win.cc',
+            'rw_lock_win.h',
+            'thread_win.cc',
+            'thread_win.h',
+            'trace_win.cc',
+            'trace_win.h',
+          ],
         }],
         ['OS=="mac"', {
           'link_settings': {
